ostd/src/task/processor.rs

```diff
@@ -96,16 +90,60 @@ pub(super) fn switch_to_task(next_task: Arc<Task>) {
         context_switch(current_task_ctx_ptr, next_task_ctx_ptr);
     }
 
-    // Now it's fine to drop `prev_task`. However, we choose not to do this because it is not
-    // always possible. For example, `context_switch` can switch directly to the entry point of the
-    // next task. Not dropping is just fine because the only consequence is that we delay the drop
-    // to the next task switching.
-
-    // See also `kernel_task_entry`.
-    crate::arch::irq::enable_local();
+    on_task_entry_or_resume();
 
-    // The `next_task` was moved into `CURRENT_TASK_PTR` above, now restore its FPU state.
     if let Some(current) = Task::current() {
         current.restore_fpu_state();
     }
 }
+
+fn update_prev_task_ptr(current_task_ptr: *const Task) {
+    #[cfg(debug_assertions)]
+    {
+        let old_prev = PREVIOUS_TASK_PTR.load();
+        debug_assert!(old_prev.is_null());
```

Comment:
I recently learned that the statement in `debug_assert!` does not run at all on release builds, see [the doc](https://doc.rust-lang.org/std/macro.debug_assert.html) and [the code](https://doc.rust-lang.org/src/core/macros/mod.rs.html#332).
> An optimized build will not execute debug_assert! statements unless -C debug-assertions is passed to the compiler. This makes debug_assert! useful for checks that are too expensive to be present in a release build but may be helpful during development.

This is indeed surprising to me (I thought the statement is executed but the result is ignored before). But if so, we can simplify the code as `debug_assert!(PREVIOUS_TASK_PTR.load().is_null())`.